### ISSUE TITLE
Mnemonic-LM: Train a bespoke local model to replace cloud LLM dependencies

### ISSUE THREAD

**Issue by CalebisGross** [AppSprout-dev/mnemonic#149]
# Mnemonic-LM

Train a purpose-built local LLM using Felix-LM v3 (hub-and-spoke architecture) and embed it directly in the mnemonic binary. No cloud APIs. No external inference server. One binary, one model file, one database.

A model trained on cognitive neuroscience literature, powering a system built on cognitive neuroscience principles. It doesn't just output JSON — it understands memory.

## The Vision

```
Today:    mnemonic (Go) → HTTP → Gemini/LM Studio (external) → response
Goal:     mnemonic (Go + embedded model) → direct inference → response
```

Fully self-contained: `~/.mnemonic/models/felix-encoder-v1.gguf` loaded at startup via CGo + llama.cpp. Install mnemonic, and it just works. Self-improving — the model gets smarter every night while you sleep.

## Why Felix-LM v3

Caleb's own architecture. Hub-and-spoke design: standard transformer backbone with lightweight low-rank spoke probes per layer. Already beating baseline transformers at 100M params (39.15 PPL, -2.7%). 104 autoresearch experiments completed. Proven on the target hardware (RX 7800 XT, ROCm).

Spokes can specialize per encoding subtask (factual extraction, semantic analysis, classification, salience calibration).

## Architecture

| Model | Task | Size | Architecture |
|-------|------|------|-------------|
| **Encoder** | Encoding, abstraction, perception (80%+ of calls) | 100M → 500M | Felix-LM v3, embedded via llama.cpp |
| **Synthesizer** | Retrieval synthesis with tool-use | 100M → 500M | Felix-LM v3, embedded via llama.cpp |
| **Embedder** | Vector embeddings | 100-300M | Embedded (ONNX or llama.cpp) |

---

## Pretraining Data: Domain-Specific, Not Generic Web Crawl

The data mix is the DNA of the model. Mnemonic implements cognitive science algorithms — encoding (levels-of-processing theory), consolidation (sleep memory research), spread activation retrieval (Collins & Loftus), episodic memory (Tulving), salience (Knudsen), metacognition (Nelson & Narens). A model trained on this literature will intrinsically understand the processes it powers.

**Custom BPE tokenizer** (32K vocab) trained on our data mix — optimized for neuroscience terms, Go/Python code, and JSON syntax. Not GPT-2.

### Data Mix (~6.5B tokens collected)

| Source | Weight | Tokens | Status |
|--------|--------|--------|--------|
| Cognitive neuro / psych papers (PeS2o) | 22% | 1.63B | ✅ Done |
| CS papers (PeS2o) | 10% | 48M | ✅ Done |
| Code (StarCoderData) | 28% | 561M | ✅ Done |
| FineWeb-Edu | 18% | 4.19B | ✅ Done |
| StackOverflow | 8% | 35M | ✅ Done |
| JSON / structured data | 5% | 29M | ✅ Done |
| Git commits + changelogs | 5% | 28M | ✅ Done |

### Data Sources (all legal, open access)

- [PeS2o](https://huggingface.co/datasets/allenai/peS2o) — 40M open-access papers, filtered by field + keywords
- [StarCoderData](https://huggingface.co/datasets/bigcode/starcoderdata) — Deduplicated code (Go, Python, JS, bash, SQL)
- [FineWeb-Edu](https://huggingface.co/datasets/HuggingFaceFW/fineweb-edu) — Highest-quality filtered web text
- [CommitPackFT](https://huggingface.co/datasets/bigcode/commitpackft) — Curated commit messages
- StackOverflow data dumps — Technical Q&A
- SchemaStore, JSONSchemaBench — JSON schemas + structured data

---

### Phase 0: Validation & Baselines
- [x] Run mnemonic benchmark suite against current Gemini (quality floor) — #173, PR #178
- [ ] Test raw Qwen 3.5 4B against mnemonic encoding task (comparison)
- [x] Profile Felix-LM v3 VRAM at 500M on 7800 XT — **500M fits at batch 4 (7GB) but pretraining at 6.5B tokens would take ~64 days. 100M is the practical local training target.**
- [x] Verify GGUF export path for Felix-LM v3 custom architecture — custom "felix" arch tag, 383 tensors (183 hub + 200 spoke), F16/F32 mixed

### Phase 1: Training Data ✅ — Milestone: [Phase 1: Pretraining Data Pipeline](https://github.com/AppSprout-dev/mnemonic/milestone/3)

**Fine-tuning data (mnemonic-specific):**
- [x] Build training data capture pipeline — #148
- [x] Build quality gate pipeline — #148
- [ ] Accumulate >5K validated encoding examples (running)

**Pretraining data (general knowledge base):**
- [x] Train custom BPE tokenizer (32K vocab) — #150
- [x] Download & filter PeS2o academic papers (484K neuro, 200K CS) — #151
- [x] Download & filter StarCoderData (500K docs, 6 languages) — #152
- [x] Download & filter FineWeb-Edu (4.1M docs) — #153
- [x] Download & parse StackOverflow (200K posts) — #154
- [x] Download JSON/structured data + git commits — #155
- [x] Build MixedPretrainDataset + test training integration — #156
- [x] Validate pipeline: ratios, domain coverage, sequence quality (audit_mix.py)
- [x] Training bridge to Felix-LM v3 (smoke test passed: loss 9.54→6.96 in 1000 steps)

**Corpus: 6.53B tokens, 7 sources, 69 shards, 13GB tokenized. Custom 32K BPE tokenizer.**

### Phase 2: Felix-LM v3 Pretraining
- [x] Autoresearch: optimize pretraining hyperparameters (LR, wd, beta2, warmup, batch size) via short sweep runs — EXP-1/2/3, LR 3.5e-3 optimal
- [x] Pretrain 100M encoder on curated data mix (local, 7800 XT) — 34.4h, loss 2.512, PPL 12.3, 100K micro-steps
- [ ] Pretrain 500M encoder (cloud GPU or extended local run)
- [ ] Scale to 1B if 500M shows promise

### Phase 3: Task-Specific Fine-Tuning
- [x] Fine-tune for encoding (JSON schema, controlled vocabulary) — 0.157 BPP, 3304 examples, spoke gates 0.08-0.99
- [x] Fine-tune for synthesis (tool-use, multi-turn) — EXP-6/EXP-9: 195 synthetic examples, mixed fine-tune, eval loss 0.522
- [x] Fine-tune embedding model on mnemonic domain — EXP-7: nomic-v2-moe fine-tuned on 50K triplets, nDCG +76.8% over base
- [x] Autoresearch: optimize spoke specialization per encoding subtask — EXP-8: spokes specialize by depth (0.08-0.99) but not by subtask; router needed

### Phase 4: GGUF Export & Embedded Inference
- [x] Build GGUF exporter for Felix-LM v3 (custom spoke layers) — export_gguf.py, 235 MB F16, RoPE absorption + embed_proj fusion
- [x] Integrate llama.cpp into mnemonic via CGo — appsprout-dev/llama.cpp fork (felix branch), 263 lines C++, 20 MB binary
- [x] Implement `EmbeddedProvider` (new `llm.Provider`, in-process inference) — #174, PR #180
- [x] Constrained decoding via llama.cpp grammar sampling — #174, PR #180
- [x] Model file management (`~/.mnemonic/models/`, versioning) — #174, PR #180

### Phase 5: Autoresearch Evolution Loop
- [ ] Composite mnemonic quality score (JSON compliance, field quality, vocab alignment, retrieval NDCG)
- [ ] Overnight experiments (~100/night)
- [ ] Spoke specialization: do spokes develop distinct encoding circuits?

### Phase 6: Integration & Cutover
- [x] Config: `llm.provider: "embedded"` vs `"api"` (backwards compatible) — #174, PR #180
- [ ] Shadow mode A/B testing (embedded vs Gemini)
- [ ] Task-by-task cutover: perception → abstraction → encoding → synthesis
- [ ] Ship model with releases or `mnemonic download-model`

### Phase 7: Self-Improving Loop
- [ ] Capture → quality gates → retrain → export GGUF → hot-reload
- [ ] `mnemonic upgrade-model` CLI command
- [ ] The flywheel: better model → better memories → better training data → better model

---

## Hardware

- AMD RX 7800 XT (16GB VRAM, ROCm) — practical for 100M pretraining, 500M fine-tuning
- Secondary: Mac Mini M4

## Related PRs & Issues

- #148 — Training data capture pipeline (merged)
- #150 — Custom BPE tokenizer (closed)
- #151 — PeS2o academic papers (closed)
- #152 — StarCoderData (closed)
- #153 — FineWeb-Edu (closed)
- #154 — StackOverflow (closed)
- #155 — JSON + git commits (closed)
- #156 — MixedPretrainDataset + training integration (closed)
- Felix-LM v3: `~/Projects/felixlm`

**Comment by CalebisGross** [comment]
## Pivoting from Felix-LM 100M to Qwen 3.5 + Spoke Adapters

### What happened with Felix 100M
Phase 3-4 work is merged (#360). The CGo llama.cpp backend, GGUF export, and fine-tuning pipeline all work. But the 100M model itself can't generalize:

- **EXP-9** (3K examples, 3 epochs): Produces valid JSON but hallucinates the same template on novel inputs
- **EXP-10** (14K examples, 5 epochs): Degenerate repetitive output, can't even produce valid JSON on novel inputs. Root cause: LR was 3.5e-3 (pretrain-level) instead of ~3.5e-5 for fine-tuning — catastrophic forgetting

The 100M scale is capacity-limited for this task. Training from scratch was also expensive (34h pretrain + 20h fine-tune on a single RX 7800 XT).

### New path: Qwen 3.5 + Felix spokes
Start from a pretrained Qwen 3.5 base (0.8B or 2B) and add Felix spoke adapters on top. This gives us:

- **Free language/JSON capability** — Qwen already knows English, JSON, structured output
- **Spoke routing** — lightweight low-rank probes for task-specific adaptation
- **Fast fine-tuning** — hours not days, spoke-only training with frozen base
- **Standard GGUF** — Qwen is already supported in llama.cpp

EXP-11 through EXP-14 are pre-registered for the Qwen spoke work. Infrastructure merged in #359. Handoff doc at `training/docs/handoff_qwen_spoke_integration.md`.

Also exploring Google's TurboQuant (March 25) for KV cache compression at inference time — training-free, 3-4 bit KV cache, could significantly reduce memory footprint in the daemon.
